### ISSUE TITLE
Text in Hero glitches during transition

### ISSUE THREAD

**Issue by grabbou** [flutter/flutter#12463]
## Steps to Reproduce

When animating between two pages that have `Hero` that wraps `Text`, the `Text` component will loose its styling for the in-transition state unless you explicitly provide it with `fontFamily`, `textDecoration` and `color`. Start and end is okay. I believe this could be related to the way we interpolate the style for the default values.

Here's the code that glitches:
```dart
var title = new Hero(
      tag: "${category.id}-title",
      child: new Text(
        category.name,
        style: const TextStyle(
          fontSize: 30.0,
        ),
        textAlign: TextAlign.center,
      ),
    );
```
<img src="https://user-images.githubusercontent.com/2464966/31335274-1e5ea28a-acf2-11e7-8283-3ca2a24a3563.gif" height="400" />

and here's the one, that is not:
```dart
var title = new Hero(
      tag: "${category.id}-title",
      child: new Text(
        category.name,
        style: const TextStyle(
          fontSize: 30.0,
          fontWeight: FontWeight.normal,
          fontFamily: "Roboto",
          color: Colors.black,
          decoration: TextDecoration.none,
        ),
        textAlign: TextAlign.center,
      ),
    );
```

Note that I have explicitly overwritten properties that were lacking proper values during transition. Is that intentional?

## Flutter Doctor

```
[✓] Flutter (on Mac OS X 10.12.6 16G29, locale en-PL, channel alpha)
    • Flutter at /Users/grabbou/Repositories/flutter
    • Framework revision 701d534ba9 (4 weeks ago), 2017-09-12 14:01:51 -0700
    • Engine revision 31d03de019
    • Tools Dart version 1.25.0-dev.11.0

[✓] Android toolchain - develop for Android devices (Android SDK 25.0.2)
    • Android SDK at /usr/local/Cellar/android-sdk
    • Platform android-25, build-tools 25.0.2
    • ANDROID_HOME = /usr/local/Cellar/android-sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_112-release-b06)

[-] iOS toolchain - develop for iOS devices (Xcode 9.0)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 9.0, Build version 9A235
    ✗ libimobiledevice and ideviceinstaller are not installed or require updating. To update, run:
        brew uninstall --ignore-dependencies libimobiledevice
        brew install --HEAD libimobiledevice
        brew install ideviceinstaller
    ✗ ios-deploy not installed. To install:
        brew install ios-deploy
    • CocoaPods version 1.2.0

[✓] Android Studio
    • Android Studio at /Applications/Android Studio 3.0 Preview.app/Contents
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-915-b01)

[✓] Android Studio (version 2.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Java version OpenJDK Runtime Environment (build 1.8.0_112-release-b06)

[✓] Connected devices
    • Custom Phone, 6 0 0, API 23, 768x1280 • 192.168.56.101:5555 • android-x86 • Android 6.0 (API 23)
```

**Comment by eseidelGoogle** [comment]
@Hixie or @HansMuller would know.

**Comment by grabbou** [comment]
I've talked to @Hixie on Gitter and we have confirmed this to be an issue. 

My understanding is that we can either fix this immediately - fix should be more or less similar to the recent gradient works. I would be happy to land it, but would require some assistance if possible.

Alternatively, we could submit PR mentioning this issue in documentation. I believe it would be valuable to be aware of it. It took me a while to figure out the root cause of that problem.

**Comment by Hixie** [comment]
The problem is how to figure out which Inherited widgets to inherit from and lerp, and how to get their status when the hero is building (which typically happens long before either the start or end of the animation builds, if they have to rebuild). The workaround is to just provide a DefaultTextStyle in the hero itself.

**Comment by Hixie** [comment]
See https://github.com/flutter/flutter/issues/213 for our last (failed) attempt to fix this.

**Comment by grabbou** [comment]
Thanks for linking to the issue. That's what I have came up with so far. Shall we mention that in a documentation?

**Comment by renefloor** [comment]
I just wanted to use a hero transition with a text going from 'subhead' to 'title'. It moves, but the layout doesn't animate nicely. Seeing #213 I have little hope that this will ever be made. What are the chances that texts in heros will animate nicely somewhere in the coming months?

**Comment by Hixie** [comment]
I don't expect that we'll attempt to resolve this again in the coming months, though it is definitely something we want to eventually provide a solution for.

We would certainly be open to reviewing patches that attempt to address this issue.

Note that animating text sizes and styles in general is not something that will be smooth even in the medium term, because fonts change shape as they change size, so just animating the font size alone looks quite ugly. We usually work around this by leaving the text at one font size but animating a transform (scale) of the text instead. This usually gives a much nicer effect.

**Comment by renefloor** [comment]
Yes I understand that. I'll see what I can do with your tips

**Comment by AlexeySemigradsky** [comment]
Any news?

**Comment by Hixie** [comment]
No change since my last comment. It's on our list of things to fix, but is not on our list of things to fix soon.

**Comment by littlemarc2011** [comment]
It's actually funny because I was trying to reproduce the same design as can be seen here:
![todo app](https://user-images.githubusercontent.com/13179783/39776454-8527f712-52c6-11e8-95b1-dacd08049e96.gif)
Anyhow, you have to wrap the text in a Material. Example:
```
new Hero(
  tag: todoObject.uuid + "_title",
  child: new Material(
    color: Colors.transparent,
    child: new Text(
      todoObject.title,
      style: new TextStyle(
        fontSize: 30.0
      ),
    ),
  ),
)
```

Hope this helps!
-Marc

**Comment by reeichert** [comment]
Any news?

**Comment by RazaGR** [comment]
any update on this?

**Comment by dukaric1991** [comment]
any news :)?

**Comment by dukaric1991** [comment]
littlemarc2011's solution works :)

**Comment by ozexpert** [comment]
@littlemarc2011 what is the icon used in your app for that menu button?

**Comment by littlemarc2011** [comment]
> @littlemarc2011 what is the icon used in your app for that menu button?

[Here](https://github.com/littlemarc2011/FlutterTodo) is the source, but if I remember correctly from a while ago it is a custom icon/font.

**Comment by modulovalue** [comment]
This did the trick for me
```dart
return Hero(
    flightShuttleBuilder: (_,
        Animation<double> animation,
        HeroFlightDirection flightDirection,
        BuildContext fromHeroContext,
        BuildContext toHeroContext) {
        return AnimatedBuilder(
            animation: animation,
            child: child,
            builder: (_, _child) {
                return DefaultTextStyle.merge(
                    child: _child,
                    style: TextStyle.lerp(DefaultTextStyle
                        .of(fromHeroContext)
                        .style, DefaultTextStyle
                        .of(toHeroContext)
                        .style, flightDirection == HeroFlightDirection.pop ? 1 - animation.value :
                                animation.value),
                );
            },
        );
    },
    tag: _tag,
    child: child,
);
```

**Comment by LongCatIsLooong** [comment]
Dup of #36220. This issue is older but since I assigned #36220 to myself I'm going to close this instead. This should be fixed once #36220 is closed.

**Comment by zyllus17** [comment]
This is the best solution provided by @HansMuller  I found among all the solutions.

Create this widget -->
```
Widget flightShuttleBuilder(
    BuildContext flightContext,
    Animation<double> animation,
    HeroFlightDirection flightDirection,
    BuildContext fromHeroContext,
    BuildContext toHeroContext,
    ) {
  return DefaultTextStyle(
    style: DefaultTextStyle.of(toHeroContext).style,
    child: toHeroContext.widget,
  );
}
```
Call the widget after tag of every Hero() like this in both the places where you have used Hero animation.
```
child: Hero(
          tag: 'Home',
          flightShuttleBuilder: flightShuttleBuilder,
          child: Expanded(
```
Unlike @HansMuller widget keep it public so you can call it anywhere, if you keep it private, then you have to create the widget in every page.
